storage/innobase/xtrabackup/src/redo_log.cc

```diff
@@ -776,6 +785,51 @@ void Archived_Redo_Log_Monitor::thread_func() {
       hdr_len -= n_read;
     }
 
+    /* for 8.0.22 onwards first block of redo archive file is all zeros */
+    if (log_block_get_checksum(buf) == 0) {
+      archive_first_block_zero = true;
+      hdr_len = QUEUE_BLOCK_SIZE - OS_FILE_LOG_BLOCK_SIZE;
+      aligned_array_pointer<byte, UNIV_PAGE_SIZE_MAX> buf2;
+      buf2.create(QUEUE_BLOCK_SIZE - OS_FILE_LOG_BLOCK_SIZE);
+      while (hdr_len > 0 && !stopped) {
+        size_t n_read = my_read(file, buf2, hdr_len, MYF(MY_WME));
+        if (n_read == MY_FILE_ERROR) {
+          msg("xtrabackup: cannot read from '%s'\n", archive.filename.c_str());
+          mysql_close(mysql);
+          my_thread_end();
+          return;
+        }
+        os_event_wait_time_low(event, 100 * 1000, 0);
+        os_event_reset(event);
+        hdr_len -= n_read;
+      }
+
+      hdr_len = OS_FILE_LOG_BLOCK_SIZE;
+      while (hdr_len > 0 && !stopped) {
+        size_t n_read = my_read(file, buf, hdr_len, MYF(MY_WME));
+        if (n_read == MY_FILE_ERROR) {
+          msg("xtrabackup: cannot read from '%s'\n", archive.filename.c_str());
+          mysql_close(mysql);
+          my_thread_end();
+          return;
+        }
+        os_event_wait_time_low(event, 100 * 1000, 0);
+        os_event_reset(event);
+        hdr_len -= n_read;
+      }
+      if (srv_redo_log_encrypt) {
+        aligned_array_pointer<byte, OS_FILE_LOG_BLOCK_SIZE> scratch_buf;
+        scratch_buf.create(OS_FILE_LOG_BLOCK_SIZE);
```

Comment:
How is the scratch_buf freed?

Reply:
(gdb) bt
#0  ut_allocator<unsigned char>::deallocate (this=0x7ffff0035930, ptr=0x7fffe80086b0 "", n_elements=0) at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:722
#1  0x00005555585aa686 in aligned_memory<unsigned char, 512ul>::free_memory (this=0x7ffff0035a40) at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:1215
#2  0x00005555585aa08d in aligned_memory<unsigned char, 512ul>::~aligned_memory (this=0x7ffff0035a40, __in_chrg=<optimized out>)
    at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:1173
#3  0x00005555585abeb8 in aligned_array_pointer<unsigned char, 512ul>::~aligned_array_pointer (this=0x7ffff0035a40, __in_chrg=<optimized out>)
    at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:1260
#4  0x00005555585a65e4 in Archived_Redo_Log_Monitor::thread_func (this=0x7fffffffd8b0) at /home/rahul.malik/MySQL/src/x28/storage/innobase/xtrabackup/src/redo_log.cc:822
#5  0x00005555585a54f7 in Archived_Redo_Log_Monitor::<lambda()>::operator()(void) const (__closure=0x7ffff0035ce0)
    at /home/rahul.malik/MySQL/src/x28/storage/innobase/xtrabackup/src/redo_log.cc:586
#6  0x00005555585a92bb in std::__invoke_impl<void, Archived_Redo_Log_Monitor::start()::<lambda()>&>(std::__invoke_other, Archived_Redo_Log_Monitor::<lambda()> &) (__f=...)
    at /usr/include/c++/7/bits/invoke.h:60